Initial precompile configurations (#1026)
ava-labs/avalanche-docs: docs/subnets/customize-a-subnet.md

### Restricting Smart Contract Deployers

If you'd like to restrict who has the ability to deploy contracts on your
subnet, you can provide an `AllowList` configuration in your genesis or upgrade file:

```json
{
  "contractDeployerAllowListConfig": {
    "blockTimestamp": 0,
    "adminAddresses": ["0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"]
  }
}
```

In this example, `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` is named as the
`Admin` of the `ContractDeployerAllowList`. This enables it to add other `Admin` or to add
`Enabled` addresses. Both `Admin` and `Enabled` can deploy contracts. To provide
a great UX with factory contracts, the `tx.Origin` is checked for being a valid
deployer instead of the caller of `CREATE`. This means that factory contracts will still be
able to create new contracts as long as the sender of the original transaction is an allow
listed deployer.

The `Stateful Precompile` contract powering the `ContractDeployerAllowList` adheres to the [AllowList Solidity interface](#allowlist-interface) at `0x0200000000000000000000000000000000000000` (you can load this interface and interact directly in Remix):

> If you attempt to add a `Enabled` and you are not an `Admin`, you will see

Reviewer: ```suggestion
- If you attempt to add a `Enabled` and you are not an `Admin`, you will see
```
Should this `If` and the other 2 ifs be under this `:`?

Reviewer 2: This should be the IAllowList here: https://github.com/ava-labs/subnet-evm/blob/master/contract-examples/contracts/IAllowList.sol

Reviewer: Not sure I understand what needs to be done with this.

Reviewer 2: We should copy and paste this interface so that it goes under the section following `... directly in Remix):` and shows the user exactly the interface that they can use in Remix.

There is another example of what this looks like for the solidity interfaces below

Reviewer: We just had that interface in line 234-251 and we should not duplicate it here.

something like:
![admin fail](/img/admin_fail.png)

> If you attempt to deploy a contract but you are not an `Admin` not

Reviewer: ```suggestion
- If you attempt to deploy a contract but you are not an `Admin` not
```

a `Enabled`, you will see something like:
![deploy fail](/img/deploy_fail.png)

> If you call `readAllowList(addr)` then you can read the current role of `addr`, which will return a uint256 with a value of 0, 1, or 2, corresponding to the roles `None`, `Enabled`, and `Admin` respectively.

Reviewer: ```suggestion
- If you call `readAllowList(addr)` then you can read the current role of `addr`, which will return a uint256 with a value of 0, 1, or 2, corresponding to the roles `None`, `Enabled`, and `Admin` respectively.
```

:::warning

If you remove all of the admins from the allow list, it will no longer be possible to update the allow list without modifying the subnet-evm to schedule a network upgrade.

:::

#### Initial Configuration

It's possible to enable this precompile with an initial configuration to activate it's effect on activation timestamp. This provides a way to enable the precompile without an admin address to manage the deployer list. With this, you can define a list of addresses that are allowed to deploy contracts. Since there will be no admin address to manage the deployer list, it can only be modified through a network upgrade. To use initial configuration, you need to specify addresses in `enabledAddresses` field in your genesis or upgrade file:

```json
{
  "contractDeployerAllowListConfig": {
    "blockTimestamp": 0,
    "enabledAddresses": ["0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"]
  }
}
```

> This will allow only `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` to deploy contracts. For further information about precompile initial configurations see [Initial Configs](#initial-configs).

Reviewer: where should #initial-configs link to?

Reviewer 2: There's no intiial configs section right now, let's just remove this sentence

Reviewer: k

    "adminAddresses": [<list of addresses>],
    "enabledAddresses": [<list of addresses>]
  }
}
```

`AllowList` configuration affects only the related precompile. For instance, the admin address in `feeManagerConfig` does not affect admin addresses in other activated precompiles.

The `AllowList` solidity interface is defined as follows:

```solidity
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

interface IAllowList {
  // Set [addr] to have the admin role over the precompile
  function setAdmin(address addr) external;

  // Set [addr] to be enabled on the precompile contract.
  function setEnabled(address addr) external;

  // Set [addr] to have no role the precompile contract.
  function setNone(address addr) external;

  // Read the status of [addr].
  function readAllowList(address addr) external view returns (uint256 role);
}
```

`readAllowList(addr)` will return a uint256 with a value of 0, 1, or 2, corresponding to the roles `None`, `Enabled`, and `Admin` respectively.

_Note: `AllowList` is not an actual contract but just an interface. It's not callable by itself. This is used by other precompiles. Check other precompile sections to see how this works._

> ### Restricting Smart Contract Deployers

Reviewer: ```suggestion
#### Restricting Smart Contract Deployers
```
so that it is under the header `AllowList`?

Author: no. they should be equal.

WARNING: if you remove all of the admins from the allow list, it will no longer be possible to update the allow list without modifying the subnet-evm to schedule a network upgrade.

#### Initial Configuration

It's possible to enable this precompile with an initial configuration to activate it's effect on activation timestamp. This provides a way to enable the precompile without an admin address to manage the deployer list. With this, you can define a list of addresses that are allowed to deploy contracts. Since there will be no admin address to manage the deployer list, it can only be modified through a network upgrade. To use initial configuration, you need to specify addresses in `enabledAddresses` field in your genesis or upgrade file:

```json
{
  "contractDeployerAllowListConfig": {
    "blockTimestamp": 0,
    "enabledAddresses": ["0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"]
  }
}
```

This will allow only `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` to deploy contracts. For further information about precompile initial configurations see [Initial Configs](#initial-configs).

> ### Restricting Who Can Submit Transactions

Reviewer: ```suggestion
#### Restricting Who Can Submit Transactions
```

Author: all precompiles use allowlist. so grouping only these two with allowlist might be confusing.

WARNING: if you remove all of the admins from the allow list, it will no longer be possible to update the allow list without modifying the subnet-evm to schedule a network upgrade.

#### Trustless Mode

> It's possible to enable this precompile without an admin address. In this mode you can define a list of addresses that are allowed to deploy contracts. This list is immutable and can't be changed. This mode is useful for networks that don't need a trusted party to manage the allow list. To enable this mode, you need to specify addresses in `enabledAddresses` field in your genesis or upgrade file:

Reviewer 2: Can we replace the blurb `This list is immutable and can't be changed` with `This list requires an admin address to modify it or a network upgrade. If there are no admin addresses, this list can only be modified through a network upgrade` throughout this file.

The allow list has three roles: `None`, `Enabled`, and `Admin`.

If you call `readAllowList(addr)` then you can read the current role of `addr`, which will return a `uint256` with a value of 0, 1, or 2, corresponding to the roles `None`, `Allowed`, and `Admin` respectively.

:::warning

If you remove all of the admins from the allow list, it will no longer be possible to update the allow list without modifying the subnet-evm to schedule a network upgrade.

:::

#### Initial Configuration

It's possible to enable this precompile with an initial configuration to activate it's effect on activation timestamp. This provides a way to enable the precompile without an admin address to manage the tx allow list. With this, you can define a list of addresses that are allowed to submit transactions. Since there will be no admin address to manage the tx list, it can only be modified through a network upgrade. To use initial configuration, you need to specify addresses in `enabledAddresses` field in your genesis or upgrade file:

```json
{
  "txAllowListConfig": {
    "blockTimestamp": 0,
    "enabledAddresses": ["0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"]
  }
}
```

> This will allow only `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` to submit transactions. For further information about precompile initial configurations see [Initial Configs](#initial-configs).

Reviewer: #initial-configs is an invalid link.

Reviewer 2: Same here

WARNING: if you remove all of the admins from the allow list, it will no longer be possible to update the allow list without modifying the subnet-evm to schedule a network upgrade.

> #### Trustless Mode

Reviewer 2: We should rename `Trustless Mode` and just document this as there are admin addresses with the ability to modify the allow list, allow listed addresses with the privilege to modify the fee config, mint native coin, etc. and the initial configuration which sets the value when the precompile activates.

}
```

#### Trustless Mode

> It's possible to enable this precompile without an admin address. In this mode you can define a list of addresses that will receive native coins. This mode is useful for networks that requires one-time mint only without a trusted party. To enable this mode, you need to specify a map of addresses with their corresponding mint amounts in `initialMint` field in your genesis or upgrade file:

Reviewer 2: ```suggestion
It's possible to enable this precompile without an admin address. In this mode you can define a list of addresses that will receive an initial mint of the native coin when this precompile activates. This mode is useful for networks that require a one-time mint without specifying any allow listed or admin addresses with privileges to mint the native coin. To enable this mode, you need to specify a map of addresses with their corresponding mint amounts in `initialMint` field in your genesis or upgrade file:
```

```json
{
  "contractNativeMinterConfig": {
    "blockTimestamp": 0,
    "initialMint": {
      "0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC": "1000000000000000000",
      "0x10037Fb06Ec4aB8c870a92AE3f00cD58e5D484b3": "0xde0b6b3a7640000"
    }
  }
}
```

> In the amount field you can specify either decimal or hex string. This will mint 1000000000000000000 (equivalent of 1 Coin) to both addresses. Note that these are both in string format. "0xde0b6b3a7640000" hex is equivalent to 1000000000000000000. For further information about trustless precompiles see [Trustless Precompile Activation](#trustless-precompile-activation).

Reviewer 2: ```suggestion
In the amount field you can specify either decimal or hex string. This will mint 1000000000000000000 (equivalent of 1 Native Coin denominated as 10^18) to both addresses. Note that these are both in string format. "0xde0b6b3a7640000" hex is equivalent to 1000000000000000000. For further information about trustless precompiles see [Trustless Precompile Activation](#trustless-precompile-activation).
```

“:100000000,\“baseFeeChangeDenominator\“:48,\“minBlockGasCost\“:0,\“maxBlockGasCost\
“:10000000,\“blockGasCostStep\“:500000}, AllowFeeRecipients: false, NetworkUpgrades: {\
“subnetEVMTimestamp\“:0}, PrecompileUpgrade: {}, UpgradeConfig: {\"precompileUpgrades\":[{\"feeManagerConfig\":{\"adminAddresses\":[\"0x8db97c7cece249c2b98bdc0226cc4c2a57bf52fc\"],\"enabledAddresses\":null,\"blockTimestamp\":1668950000}},{\"txAllowListConfig\":{\"adminAddresses\":[\"0x8db97c7cece249c2b98bdc0226cc4c2a57bf52fc\"],\"enabledAddresses\":null,\"blockTimestamp\":1668960000}},{\"feeManagerConfig\":{\"adminAddresses\":null,\"enabledAddresses\":null,\"blockTimestamp\":1668970000,\"disable\":true}}]}, Engine: Dummy Consensus Engine}"”
```

You can also check activation of precompiles with [`getActivatePrecompilesAt API`](#getActivatePrecompilesAt) and the current chain config with [`getChainConfig API`](#eth_getChainConfig).

Notice that `precompileUpgrades` entry correctly reflects the changes. That's it, your Subnet is all set and the desired upgrades will be activated at the indicated timestamp!

### Trustless Precompile Activation

> Precompiles are contracts that has certain features. These features could only be triggered by privileged addresses in previous releases of subnet-evm. For example, the `feeManagerConfig` precompile allows privileged addresses to change the fee structure of the network:

Reviewer 2: Can we simplify this? The first sentence doesn't say much.

```json
{
  "precompileUpgrades": [
    {
      "feeManagerConfig": {
        "blockTimestamp": 1668950000,
        "adminAddresses": ["0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"]
      }
    }
  ]
}
```

> In this example, only the address `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` is allowed to change the fee structure of the network. The admin address has to intract with the precompile in order to activate it's effect. I.e it needs to send a transaction with relevant fee structure to change it. This is a very powerful feature, but it also means that the network is vulnerable to a single point of failure. If the address `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` is compromised, the network is compromised.

Reviewer 2: ```suggestion
In this example, only the address `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` is allowed to change the fee structure of the network. The admin address has to call the precompile in order to activate its effect. I.e it needs to send a transaction with a new fee config to perform the update. This is a very powerful feature, but it also gives a large amount of power to the admin address. If the address `0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC` is compromised, the network is compromised.
```

In order to make these precompiles more trustless, we have introduced a new way to activate precompiles without admin addresses. This is done by setting relevant precompile initial configuration in either as an upgrade or as a genesis configuration. For example, the `feeManagerConfig` precompile can be configured as follows to change the fee structure on the activation:

```json
{
  "precompileUpgrades": [
    {
      "feeManagerConfig": {
        "blockTimestamp": 1668950000,
        "initialFeeConfig": {
          "gasLimit": 20000000,
          "targetBlockRate": 2,
          "minBaseFee": 1000000000,
          "targetGas": 100000000,
          "baseFeeChangeDenominator": 48,
          "minBlockGasCost": 0,
          "maxBlockGasCost": 10000000,
          "blockGasCostStep": 500000
        }
      }
    }
  ]
}
```

> Notice that there is no `adminAddresses` field in the configuration. This means that the precompile will be activated with the given configuration without any admin address. It also means there is no need to send a transaction to change the fee structure. The precompile will change the fee at the given timestamp.

Reviewer 2: ```suggestion
Notice that there is no `adminAddresses` field in the configuration. This means that there will be no admin address with privileges to perform another update. The precompile will simply update the fee configuration to the specified fee config when it activates.
```

This is a trustless way to activate precompiles.

> _Note: If you trustlessly activate a precompile, then it will be activated as a precompile. If you want to change the configuration, you will need to first disable it then activate the precompile again with the new configuration._

Reviewer 2: Can we simplify this? It's not clear to me what the first sentence is meant to say

“:10000000,\“blockGasCostStep\“:500000}, AllowFeeRecipients: false, NetworkUpgrades: {\
“subnetEVMTimestamp\“:0}, PrecompileUpgrade: {}, UpgradeConfig: {\"precompileUpgrades\":[{\"feeManagerConfig\":{\"adminAddresses\":[\"0x8db97c7cece249c2b98bdc0226cc4c2a57bf52fc\"],\"enabledAddresses\":null,\"blockTimestamp\":1668950000}},{\"txAllowListConfig\":{\"adminAddresses\":[\"0x8db97c7cece249c2b98bdc0226cc4c2a57bf52fc\"],\"enabledAddresses\":null,\"blockTimestamp\":1668960000}},{\"feeManagerConfig\":{\"adminAddresses\":null,\"enabledAddresses\":null,\"blockTimestamp\":1668970000,\"disable\":true}}]}, Engine: Dummy Consensus Engine}"”
```

Notice that `precompileUpgrades` entry correctly reflects the changes. You can also check the activated precompiles at a timestamp with the [`eth_getActivatePrecompilesAt`](../apis/avalanchego/apis/subnet-evm.md#ethgetactivateprecompilesat) RPC method. The [`eth_getChainConfig`](../apis/avalanchego/apis/subnet-evm.md#ethgetchainconfig) RPC method will also return the configured upgrades in the response.

That's it, your Subnet is all set and the desired upgrades will be activated at the indicated timestamp!

### Initial Precompile Configurations

> Precompiles can be managed by some priveleged addresses to change their configurations and activate their effects. For example, the `feeManagerConfig` precompile can have `adminAdresses` which can change the fee structure of the network.

Reviewer 2: Can we make it clear that the enabled address can modify the fee config and only the admin address can modify the allow list in addition to modifying the fee config?

Could we also add a blurb about how allow lists work at the top to document the separation of responsibilities from the enabled addresses and the admin addresses?
